File: ios/Classes/Mapper.swift
import Flutter
import TwilioVoice

// swiftlint:disable file_length type_body_length
public class Mapper {
    
    public static func checkStringNil(data: Any?) -> String{
        if(data == nil){
            return ""
        }else {
            return data as! String
        }
    }
    
    public static func checkBoolNil(data: Any?) -> Bool{
        if(data == nil){
            return false
        }else {
            return data as! Bool
        }
    }
    
    public static func checkDicNil(data: Any?) -> [String: Any]{
        if(data == nil){
            return ["":""]
        }else {
            return data as! [String: Any]
        }
    }
    
    public static func callToDict(_ message: Call?) -> [String: Any] {
        let emptyData =  ["data" :[
            "callSid": "",
            "isOnHold" :false,
            "to" :"",
            "from" :"",
            "isMuted" :false] as [String : Any]]
        if(message != nil){
            return [
                "data" :[
                    "callSid": checkStringNil(data: message?.sid),
                    "from" :checkStringNil(data: message?.from),
                    "to" : checkStringNil(data: message?.to),
                    "isOnHold" : checkBoolNil(data: message?.isOnHold),
                    "isMuted" :checkBoolNil(data: message?.isMuted)] as [String : Any]
            ]
        }
        return emptyData
    }
    
    public static func callInviteToDict(_ call: CallInvite?) -> [String: Any] {
        let emptyData =  ["data" :[
            "twi_call_sid" :"",
            "twi_to" :"",
            "twi_from" :"",
            "customParameters" :["":""]] as [String : Any]]
        if(call != nil){
            return [
                "data" :[
                    "twi_call_sid" : checkStringNil(data: call?.callSid),
                    "twi_from" :checkStringNil(data: call?.from),
                    "twi_to" : checkStringNil(data: call?.to),
                    "customParameters" : checkDicNil(data: call?.customParameters)
                ] as [String : Any]
            ]
        }
        return emptyData;
    }
    
    public static func cancelledCallInviteToDict(_ call: CancelledCallInvite?) -> [String: Any] {
        let emptyData =  ["data" :[
            "twi_call_sid" :"",
            "twi_to" :"",
            "twi_from" :"",
            "customParameters" :["":""],
        ] as [String : Any]]
        if(call != nil){
            return [
                "data" :[
                    "twi_call_sid" : checkStringNil(data: call?.callSid),
                    "twi_from" :checkStringNil(data: call?.from),
                    "twi_to" : checkStringNil(data: call?.to),
                    "customParameters" : checkDicNil(data: call?.customParameters)
                ] as [String : Any]
            ]
        }
        return emptyData;
    }
    
    public static func errorToDict(_ error: Error?) -> [String: Any?]? {
        if let error = error as NSError? {
            return [
                "code": error.code,
                "message": error.description
            ]
        }
        return nil
    }
}

File: android/src/main/kotlin/com/flutter/twilio/voice/Mapper.kt
package com.flutter.twilio.voice

import com.twilio.voice.*
import org.json.JSONArray
import org.json.JSONException
import org.json.JSONObject

object Mapper
{
    fun errorInfoToMap(e: RegistrationException?): Map<String, Any?>? {
        if (e == null)
            return null
        return mapOf(
                "status" to e.errorCode,
                "code" to e.errorCode,
                "message" to e.message
        )
    }

    fun errorInfoToMap(e: CallException?): Map<String, Any?>? {
        if (e == null)
            return null
        return mapOf(
                "code" to e.errorCode,
                "message" to e.message
        )
    }

    fun callToMap(message: Call): Map<String, Any?> {
        return mapOf(
                "callSid" to message.sid,
                "from" to message.from,
                "to" to message.to,
                "isOnHold" to message.isOnHold,
                "isMuted" to message.isMuted
        )
    }

    fun callInviteToMap(message: CallInvite): Map<String, Any?>
    {
        var temp: String = message.customParameters["channel_info"] as String
        temp = temp.replace("{'", "{\"");
        temp = temp.replace("':", "\":");
        temp = temp.replace(": '", ": \"");
        temp = temp.replace("',", "\",");
        temp = temp.replace(", '", ", \"");
        temp = temp.replace("'}", "\"}");
        temp = temp.replace("None", "null")
        return mapOf(
                "twi_call_sid" to message.callSid,
                "twi_to" to message.to,
                "twi_from" to message.from,
                "customParameters" to message.customParameters,
                "channelInfo" to toMap(JSONObject(temp))
        )
    }

    @Throws(JSONException::class)
    fun toMap(jsonobj: JSONObject): Map<String, Any>
    {
        val map: MutableMap<String, Any> = HashMap()
        val keys = jsonobj.keys()
        while (keys.hasNext()) {
            val key = keys.next()
            var value = jsonobj[key]
            if (value is JSONArray) {
                value = toList(value)
            } else if (value is JSONObject) {
                value = toMap(value)
            }
            map[key] = value
        }
        return map
    }

    @Throws(JSONException::class)
    fun toList(array: JSONArray): List<Any> {
        val list: MutableList<Any> = ArrayList()
        for (i in 0 until array.length()) {
            var value = array[i]
            if (value is JSONArray) {
                value = toList(value)
            } else if (value is JSONObject) {
                value = toMap(value)
            }
            list.add(value)
        }
        return list
    }

    fun cancelledCallInviteToMap(message: CancelledCallInvite): Map<String, Any?> {
        return mapOf(
                "twi_call_sid" to message.callSid,
                "twi_to" to message.to,
                "twi_from" to message.from,
                "customParameters" to message.customParameters
        )
    }
}
